crates/history/src/history.rs

```diff
@@ -0,0 +1,102 @@
+use serde::Serialize;
+
+use crate::error::HistoryResult;
+use crate::listener::HistoryListener;
+use crate::location::Location;
+
+/// A trait to provide [`History`] access.
+pub trait History: Clone + PartialEq {
+    type Location: Location<History = Self> + 'static;
+
+    /// Returns the number of elements in [`History`].
+    fn len(&self) -> usize;
+
+    /// Returns true if the current [`History`] is empty.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Moves back 1 page in [`History`].
+    fn back(&self) {
+        self.go(-1);
+    }
+
+    /// Moves forward 1 page in [`History`].
+    fn forward(&self) {
+        self.go(1);
+    }
+
+    /// Loads a specific page in [`History`] with a `delta` relative to current page.
+    ///
+    /// See: <https://developer.mozilla.org/en-US/docs/Web/API/History/go>
+    fn go(&self, delta: isize);
+
+    /// Pushes a route entry with [`None`] being the state.
+    fn push(&self, route: impl Into<String>);
+
+    /// Replaces the current history entry with provided route and [`None`] state.
+    fn replace(&self, route: impl Into<String>);
+
+    /// Pushes a route entry with state.
+    ///
+    /// The implementation of state serialization differs between [`History`] types.
+    ///
+    /// For [`BrowserHistory`] and [`HashHistory`], state is serialised with [`serde_wasm_bindgen`] where as
+    /// [`MemoryHistory`] uses [`Any`](std::any::Any).
+    fn push_with_state<T>(&self, route: impl Into<String>, state: T) -> HistoryResult<()>
+    where
+        T: Serialize + 'static;
```

Comment:
`serde` should be feature gated.

Reply:
Making serde to be optional actually will not only lose `location.state` but also `location.query`. Which I consider to be significant enough to have it on by default.

FWIW, `gloo-console` and `gloo-storage` also have serde to be a non-optional dependency.

Reply:
`gloo-storage` needs it and it should also be optional for `gloo-console`. We can feature gate but enable that feature by default.